src/blockdev.rs

```diff
@@ -99,3 +99,48 @@ pub fn find_colocated_bios_boot(devices: &Vec<String>) -> Result<Option<Vec<Stri
     log::debug!("Found bios_boot partitions: {bios_boots:?}");
     Ok(Some(bios_boots))
 }
+
+// Check if the device is mpath
+fn is_mpath(device: &str) -> Result<bool> {
+    let dm_path = Utf8PathBuf::from_path_buf(std::fs::canonicalize(device)?)
+        .map_err(|_| anyhow::anyhow!("Non-UTF8 path"))?;
+    let dm_name = dm_path.file_name().unwrap_or("");
+    let uuid_path = Utf8PathBuf::from(format!("/sys/class/block/{dm_name}/dm/uuid"));
+
+    if uuid_path.exists() {
+        let uuid = std::fs::read_to_string(&uuid_path)
+            .with_context(|| format!("Failed to read {uuid_path}"))?;
+        if uuid.trim_start().starts_with("mpath-") {
+            return Ok(true);
+        }
+    }
+    Ok(false)
+}
+
+/// Get esp partition number from device
+pub fn get_esp_partition_number(device: &str) -> Result<String> {
+    let esp_device =
+        get_esp_partition(device)?.ok_or_else(|| anyhow::anyhow!("Failed to find ESP device"))?;
+
+    if !device.starts_with("/dev/mapper") {
```

Comment:
I think it'd be cleaner here to just return the `partition` attribute if it exists, right? i.e.:

```
let partition_path = 
if partition_path.exists() {
    return read_to_string(path)
}
```